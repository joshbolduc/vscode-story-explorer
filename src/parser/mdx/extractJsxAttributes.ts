import type { NodePath } from '@babel/traverse';
import type { JSXElement } from '@babel/types';
import type { PropertyInfo } from '../PropertyInfo';
import { assignProperties } from '../csf/assignProperties';

const UNPARSED = Symbol('Unparsed value');

export const extractJsxAttributes = (path: NodePath<JSXElement>) => {
  const attributes = path.get('openingElement').get('attributes');
  const children = path.get('children');

  const mergedProps = attributes.reduce<{
    values: Record<string, PropertyInfo>;
    literals: Record<string, string>;
  }>(
    (acc, attr) => {
      if (attr.isJSXAttribute()) {
        const name = attr.node.name.name;

        if (typeof name === 'string') {
          const evalResult = attr.get('value').evaluate();
          if (evalResult.confident) {
            acc.values[name] = {
              isPartOfDeclaration: true,
              value: evalResult.value,
            };
          } else {
            acc.values[name] = {
              isPartOfDeclaration: true,
              value: UNPARSED,
            };
            const valuePath = attr.get('value');
            if (valuePath.isJSXExpressionContainer()) {
              acc.literals[name] = valuePath.get('expression').toString();
            }
          }
        }
      } else if (attr.isJSXSpreadAttribute()) {
        const evalResult = attr.get('argument').evaluate();
        const value = evalResult.value as unknown;

        if (evalResult.confident && value && typeof value === 'object') {
          assignProperties(acc.values, value);
        }
      }
      return acc;
    },
    { values: {}, literals: {} },
  );

  if (children.length > 0) {
    mergedProps.values.children = {
      isPartOfDeclaration: true,
      value: children,
    };
  }

  return mergedProps;
};
