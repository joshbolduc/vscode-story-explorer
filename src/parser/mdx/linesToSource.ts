import type { Position } from 'unist';

export const linesToSource = (lines: { value: string; position: Position }[]) =>
  lines
    .reduce<string[]>((output, { value, position }) => {
      const line = output.length;
      const desiredLine = position.start.line - 1;

      for (let i = line; i < desiredLine; i++) {
        output.push('');
      }

      output.push(...value.split('\n'));

      return output;
    }, [])
    .join('\n');
