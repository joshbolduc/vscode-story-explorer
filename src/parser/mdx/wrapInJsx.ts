const jsxPrepend = '<>';
const jsxAppend = '</>;';

export const wrapInJsx = (value: string) => `${jsxPrepend}${value}${jsxAppend}`;
