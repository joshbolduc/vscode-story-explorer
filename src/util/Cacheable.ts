/**
 * An object that lazily performs a predefined operation and caches its result.
 */
export class Cacheable<T> {
  private cachedResult?: { result: T } | undefined;

  /**
   * Initialize a new Cacheable with a given callback.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public constructor(private callback: () => T) {}

  /**
   * Get the result of the operation, invoking it if necessary. If the operation
   * has already been invoked, its result is returned.
   *
   * @returns The result of the operation, cached or otherwise.
   */
  public getResult() {
    if (!this.cachedResult) {
      this.cachedResult = { result: this.callback() };
    }

    return this.cachedResult.result;
  }

  /**
   * Get the cached result of the operation, if any, or a default value if the
   * result has not already been determined.
   */
  public getCachedResult(): T | undefined;
  public getCachedResult<U>(defaultValue: U): T | U;
  public getCachedResult<U>(defaultValue?: U) {
    return this.cachedResult ? this.cachedResult.result : defaultValue;
  }

  /**
   * Replace the callback used to computed the cached result. The cached result,
   * if any, is cleared.
   *
   * @param callback A callback that will be lazily invoked when the result is
   * requested, and whose return value will be cached and returned for
   * subsequent requests.
   */
  public setCallback(callback: () => T) {
    this.clear();
    this.callback = callback;
  }

  /**
   * Clears the cached result. The next time a result is requested, the
   * operation will be reinvoked and its return value will be cached again.
   */
  public clear() {
    this.cachedResult = undefined;
  }
}
