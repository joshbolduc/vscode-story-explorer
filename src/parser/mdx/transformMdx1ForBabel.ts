import type { ExportNode, ImportNode, JsxNode, RootNode } from '@mdx-js/mdx';
import { sync } from '@mdx-js/mdx';
import { logWarn } from '../../log/log';
import { linesToSource } from './linesToSource';
import { wrapInJsx } from './wrapInJsx';

export const transformMdx1ForBabel = (contents: string) => {
  let ast: RootNode | undefined;

  sync(contents, {
    remarkPlugins: [
      () => (tree: RootNode) => {
        ast = tree;
        return tree;
      },
    ],
  });

  if (!ast) {
    logWarn('Failed to get AST from MDX content');
    return undefined;
  }

  const lines = ast.children
    .filter((node): node is JsxNode | ImportNode | ExportNode =>
      ['jsx', 'import', 'export'].includes(node.type),
    )
    .map((node) => {
      const value = node.type === 'jsx' ? wrapInJsx(node.value) : node.value;

      return { value, position: node.position };
    });

  return linesToSource(lines);
};
