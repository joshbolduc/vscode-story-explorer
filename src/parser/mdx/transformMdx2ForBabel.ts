import { compileSync } from '@mdx-js/mdx-2';
import type { Root } from 'mdast';
import type { MdxjsEsm, MdxJsxFlowElement } from 'mdast-util-mdx';
import { logWarn } from '../../log/log';
import { linesToSource } from './linesToSource';
import { wrapInJsx } from './wrapInJsx';

export const transformMdx2ForBabel = (contents: string) => {
  let ast: Root | undefined;

  compileSync(contents, {
    remarkPlugins: [
      () => (tree: Root) => {
        ast = tree;
        return tree;
      },
    ],
  });

  if (!ast) {
    logWarn('Failed to get AST from MDX content');
    return undefined;
  }

  const lines = ast.children
    .filter((node): node is MdxjsEsm | MdxJsxFlowElement =>
      ['mdxjsEsm', 'mdxJsxFlowElement'].includes(node.type),
    )
    .map((node) => {
      if (node.type === 'mdxJsxFlowElement') {
        const value = contents.substring(
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- position expected to exist
          node.position!.start.offset!,
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- position expected to exist
          node.position!.end.offset,
        );

        return {
          value: wrapInJsx(value),
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- position expected to exist
          position: node.position!,
        };
      }

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion -- position expected to exist
      return { value: node.value, position: node.position! };
    });

  return linesToSource(lines);
};
