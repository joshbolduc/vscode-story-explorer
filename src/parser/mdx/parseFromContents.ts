import { parse } from '@babel/parser';
import traverse, { NodePath } from '@babel/traverse';
import {
  ImportDeclaration,
  isJSXIdentifier,
  JSXAttribute,
  JSXElement,
} from '@babel/types';
import type { Location } from '../../types/Location';
import type { PropertyInfo } from '../PropertyInfo';
import { sourceLocationToLocation } from '../sourceLocationToLocation';
import { extractJsxAttributes } from './extractJsxAttributes';
import { transformMdx1ForBabel } from './transformMdx1ForBabel';
import { transformMdx2ForBabel } from './transformMdx2ForBabel';

interface RawResult {
  meta: {
    location: Location | undefined;
    properties: {
      values?: Record<string, PropertyInfo>;
      literals?: Record<string, string>;
    };
    of?: { importPath: string };
    declared: boolean;
  };
  stories: {
    location: Location | undefined;
    properties: {
      values: Record<string, PropertyInfo>;
      literals: Record<string, string>;
    };
  }[];
}

const getJSXElementInfo = (path: NodePath<JSXElement>) => {
  const location = sourceLocationToLocation(path.node.loc);

  return {
    properties: extractJsxAttributes(path),
    // Fudge the location to account for preprocessing before the source gets
    // sent to babel. For now, just pretend it's zero length.
    location: location
      ? {
          start: { line: location.start.line, column: 0 },
          end: { line: location.start.line, column: 0 },
        }
      : undefined,
  };
};

const tryTransformMdxForBabel = (contents: string) => {
  try {
    return transformMdx2ForBabel(contents);
  } catch {
    // fall through
  }

  return transformMdx1ForBabel(contents);
};

export const parseFromContents = (contents: string): RawResult | undefined => {
  const result: RawResult = {
    meta: {
      location: undefined,
      properties: {},
      declared: false,
    },
    stories: [],
  };

  const combinedOutput = tryTransformMdxForBabel(contents);

  if (combinedOutput === undefined) {
    return undefined;
  }

  const parsedFile = parse(combinedOutput, {
    sourceType: 'unambiguous',
    plugins: ['jsx', 'typescript'],
    errorRecovery: true,
  });

  traverse(parsedFile, {
    JSXElement: (path) => {
      const { openingElement } = path.node;
      const { name } = openingElement;
      if (isJSXIdentifier(name)) {
        const isMeta = name.name === 'Meta';
        const isStory = name.name === 'Story';

        if (isMeta || isStory) {
          const { properties, location } = getJSXElementInfo(path);
          if (isMeta) {
            result.meta.properties = properties;
            result.meta.location = location;
            result.meta.declared = true;

            const ofAttribute = path
              .get('openingElement')
              .get('attributes')
              .find(
                (attr): attr is NodePath<JSXAttribute> =>
                  attr.isJSXAttribute() && attr.node.name.name === 'of',
              );

            const ofValue = ofAttribute?.get('value');
            if (ofValue?.isJSXExpressionContainer()) {
              const expression = ofValue.get('expression');
              if (expression.isIdentifier()) {
                const expressionName = expression.node.name;

                const binding = expression.scope.getBinding(expressionName);
                if (binding?.path.isImportNamespaceSpecifier()) {
                  const importDeclaration = binding.path.findParent((p) =>
                    p.isImportDeclaration(),
                  ) as NodePath<ImportDeclaration> | null;
                  const source = importDeclaration?.get('source');
                  const importPath = source?.node.value;

                  if (importPath) {
                    result.meta.of = { importPath };
                  }
                }
              }
            }
          } else if (isStory) {
            result.stories.push({
              location,
              properties,
            });
          }
        }
      }
    },
  });

  return result;
};
