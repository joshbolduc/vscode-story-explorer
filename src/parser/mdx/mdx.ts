import { storyNameFromExport } from '@componentdriven/csf';
import { logDebug } from '../../log/log';
import { isDefined } from '../../util/isDefined';
import type { RawParsedStoryFile } from '../RawParsedStoryFile';
import { sanitizeMetaObject } from '../sanitizeMetaObject';
import { parseFromContents } from './parseFromContents';

const parse = (contents: string): RawParsedStoryFile | undefined => {
  const parsed = parseFromContents(contents);

  if (parsed && (parsed.meta.declared || parsed.stories.length > 0)) {
    const { id, title } = sanitizeMetaObject(parsed.meta.properties.values);

    const meta = {
      id,
      title,
      location: parsed.meta.location,
    };

    const stories = parsed.stories
      .map((story) => {
        const { name: nameProp, id: idProp } = story.properties.values;

        // <Story id="..." /> just refers to another story and doesn't create an
        // entry in the store, so we'll disregard it.
        if (idProp !== undefined) {
          return undefined;
        }

        const namePropAsString =
          typeof nameProp === 'string' ? nameProp : undefined;

        const storyRefName = story.properties.literals.story;
        const storyRefShortName = storyRefName?.split('.').slice(-1)[0];
        const storyNameFromStoryRefShortName =
          storyRefShortName !== undefined
            ? storyNameFromExport(storyRefShortName)
            : undefined;

        // If not specified as a prop, the actual name will be whatever the name
        // of the imported story is. But we don't (yet) resolve imports, so
        // use `storyNameFromStoryRefShortName` as a best guess.
        const name = namePropAsString ?? storyNameFromStoryRefShortName;

        return {
          location: story.location,
          name,
          nameForId: storyNameFromStoryRefShortName,
        };
      })
      .filter(isDefined);

    return {
      meta,
      stories,
    };
  }
};

export const tryParseMdx = (contents: string) => {
  try {
    return parse(contents);
  } catch (e) {
    // Most likely due to transient syntax error
    logDebug('Failed to parse contents as MDX');
  }
  return undefined;
};
