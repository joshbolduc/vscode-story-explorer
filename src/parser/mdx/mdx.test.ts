import { readFileSync } from 'fs';
import { join } from 'path';
import { sync } from 'fast-glob';
import { describe, expect, it } from 'vitest';
import { testBaseDir } from '../../../test/util/testBaseDir';

import { tryParseMdx } from './mdx';

describe('mdx', () => {
  const tests = sync(
    [
      'project/v6/src/autoTitle/**/*.stories.mdx',
      'project/v6/src/stories/v6-only/**/*.stories.mdx',
      'project/v7/src/**/*.mdx',
      'fixtures/**/*.mdx',
    ],
    {
      cwd: testBaseDir,
    },
  );

  tests.forEach((match) => {
    it(`parses story file ${match}`, () => {
      const storiesPath = join(testBaseDir, match);
      const contents = readFileSync(storiesPath).toString();
      const result = tryParseMdx(contents);

      expect(result).toMatchSnapshot();
    });
  });
});
