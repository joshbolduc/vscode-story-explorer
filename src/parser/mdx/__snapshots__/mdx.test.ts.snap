// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mdx parses story file fixtures/Children.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": undefined,
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 1, Col 0],
      "name": "Children prop, children content",
      "nameForId": undefined,
    },
    {
      "location": [Ln 8, Col 0],
      "name": "Children prop, empty body",
      "nameForId": undefined,
    },
    {
      "location": [Ln 10, Col 0],
      "name": "Children prop, no body",
      "nameForId": undefined,
    },
    {
      "location": [Ln 12, Col 0],
      "name": "No children prop, children content",
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file fixtures/Spread.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": undefined,
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 15, Col 0],
      "name": undefined,
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file project/src/autoTitle/mdx/MdxMeta.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": undefined,
  },
  "stories": [],
}
`;

exports[`mdx parses story file project/src/autoTitle/mdx/MdxMetaAndStory.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 5, Col 0],
      "name": "Primary",
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file project/src/autoTitle/mdx/MdxStory.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": undefined,
    "title": undefined,
  },
  "stories": [
    {
      "location": [Ln 3, Col 0],
      "name": "Primary",
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file project/src/autoTitle/mdx/PlainMarkdown.stories.mdx 1`] = `undefined`;

exports[`mdx parses story file project/src/stories/CsfWithMdx.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 5, Col 0],
    "title": "Example/MDX importing CSF",
  },
  "stories": [
    {
      "location": [Ln 11, Col 0],
      "name": "Basic",
      "nameForId": "Basic",
    },
    {
      "location": [Ln 21, Col 0],
      "name": "Story name from MDX",
      "nameForId": "Basic 2",
    },
    {
      "location": [Ln 25, Col 0],
      "name": "Basic 3",
      "nameForId": "Basic 3",
    },
  ],
}
`;

exports[`mdx parses story file project/src/stories/CustomId.stories.mdx 1`] = `
{
  "meta": {
    "id": "manually-specified-mdx-id",
    "location": [Ln 3, Col 0],
    "title": "Example/Custom ID MDX",
  },
  "stories": [
    {
      "location": [Ln 7, Col 0],
      "name": "Story",
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file project/src/stories/DocsOnly.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": "Example/MDX Docs Only",
  },
  "stories": [],
}
`;

exports[`mdx parses story file project/src/stories/DocsOnlyTop.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 3, Col 0],
    "title": "MDX Docs at top level",
  },
  "stories": [],
}
`;

exports[`mdx parses story file project/src/stories/MdxWithStories.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 6, Col 0],
    "title": "Example/MDX with Stories",
  },
  "stories": [
    {
      "location": [Ln 10, Col 0],
      "name": "MDX first story",
      "nameForId": undefined,
    },
    {
      "location": [Ln 14, Col 0],
      "name": "MDX second story",
      "nameForId": undefined,
    },
    {
      "location": [Ln 18, Col 0],
      "name": "MDX Story 3",
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file project/src/stories/MdxWithStoriesTop.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 4, Col 0],
    "title": "MDX with stories, top",
  },
  "stories": [
    {
      "location": [Ln 6, Col 0],
      "name": "MDX with stories, top",
      "nameForId": undefined,
    },
  ],
}
`;

exports[`mdx parses story file project/src/stories/Precedence.stories.mdx 1`] = `
{
  "meta": {
    "id": undefined,
    "location": [Ln 4, Col 0],
    "title": "Example/Precedence",
  },
  "stories": [
    {
      "location": [Ln 20, Col 0],
      "name": "Story with name, story, body",
      "nameForId": "Small",
    },
    {
      "location": [Ln 26, Col 0],
      "name": "Story with name, body",
      "nameForId": undefined,
    },
  ],
}
`;
